pkg/kubelet/kubelet.go

```diff
@@ -2316,29 +2316,68 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
 	}
 }
 
+func readyPodCondition(isPodReady bool, reason, message string, existingStatus *api.PodStatus) []api.PodCondition {
+	currentTime := util.Now()
+	condition := api.PodCondition{
+		Type: api.PodReady,
+	}
+	if isPodReady {
+		condition.Status = api.ConditionTrue
+	} else {
+		condition.Status = api.ConditionFalse
+	}
+	condition.LastProbeTime = currentTime
+	condition.Reason = reason
+	condition.Message = message
+	if existingStatus != nil {
+		if api.IsPodReadyConditionTrue(*existingStatus) != isPodReady {
+			// condition has transitioned, update transition time.
+			condition.LastTransitionTime = currentTime
+		} else {
+			// retain the existing transition time.
+			existingCondition := api.GetPodReadyCondition(*existingStatus)
+			if existingCondition != nil {
+				condition.LastTransitionTime = existingCondition.LastTransitionTime
+			}
+		}
+	}
+	return []api.PodCondition{condition}
+}
+
 // getPodReadyCondition returns ready condition if all containers in a pod are ready, else it returns an unready condition.
-func getPodReadyCondition(spec *api.PodSpec, statuses []api.ContainerStatus) []api.PodCondition {
-	ready := []api.PodCondition{{
-		Type:   api.PodReady,
-		Status: api.ConditionTrue,
-	}}
-	unready := []api.PodCondition{{
-		Type:   api.PodReady,
-		Status: api.ConditionFalse,
-	}}
-	if statuses == nil {
-		return unready
+func getPodReadyCondition(spec *api.PodSpec, containerStatuses []api.ContainerStatus, existingStatus *api.PodStatus) []api.PodCondition {
+	// Find if all containers are ready or not.
+	if containerStatuses == nil {
+		return readyPodCondition(false, "UnknownContainerStatuses", "", existingStatus)
 	}
+	containersWithUnknownStatus := []string{}
+	containersWithUnreadyStatus := []string{}
```

Comment:
Just a nit.

I think shorter variable names here would be fine since this is not a huge function, and it's quite clearly what the code intends to do. How about  "unreadyContainers", "unknownContainers"?